docs/design.md

```diff
@@ -1,92 +1,122 @@
-+ Since the high-level workflow of processing a request is same for any request, we use the [Template Method pattern](http://en.wikipedia.org/wiki/Template_method_pattern) to abstract the process folow into the `Action` class.
-+ The list of actions and corresponding URIs are listed in the `ActionURIs` nested class of the [Const](../src/main/java/teammates/common/util/Const.java) class.
-+ The list of pages and corresponding URIs are listed in the `ViewURIs` nested class of the [Const](../src/main/java/teammates/common/util/Const.java) class.
-
-###Types of pages
-
-The UI consist of following pages:
-+ Product pages (functional): e.g., 'courses' page. These require login.
-+ Product pages (peripheral): e.g., help pages, error pages. etc.
-+ Website pages: These are the static pages of the product website. e.g., `contact.jsp`
-
-##Logic
-
-The `Logic` component handles the business logic of TEAMMATES. 
-It is accessible via a thin [facade class](http://en.wikipedia.org/wiki/Facade_pattern) called [Logic](../src/main/java/teammates/logic/api/Logic.java) which makes use of several `*Logic` classes to handle the logic related to various types of data and to access data from the `Storage` component. In particular, `Logic` is responsible for these: 
-+ Managing relationships between entities. e.g., cascade logic for create/update/delete.
-+ Managing transactions. e.g., to ensure atomicity of a transaction.
-+ Sanitizing input values recevied from the UI component.
-+ Providing a mechanism for checking access control rights.
+1. Request forwarded to the `ControllerServlet`.
+1. `ControllerServlet` uses the `ActionFactory` to generate the matching `Action` object, e.g. `InstructorHomePageAction`.
+1. `ControllerServlet` executes the action.
+1. The `Action` object checks the access rights of the user. If the action is allowed, it will be performed, interacting with the `Logic` component as necessary.
+1. The `Action` gathers the data required for the page into a `PageData` object, e.g. `InstructorHomePageData`, creates a `ShowPageResult` object by enclosing the `PageData` object created previously, and returns it to the `ControllerServlet`.
+1. `ControllerServlet` sends the result. In case of a `ShowPageResult`, this is equivalent to forwarding to the matching JSP page.
+1. The JSP page uses the data in the given `PageData` object to generate the HTML page.
+1. The response will then be sent back to the browser, which will render the page.
+
+The other 30% types of request may differ in the following manners:
+- At times, the user will request for data asynchronously via AJAX. In such cases, the actions return an `AjaxResult`. `PageData` is also created for this request and the data obtained is pre-processed for display in the browser; that is, the result of the AJAX request is not the raw data.
```

Comment:
I am aware of that. When JSON is brought into picture, the result indeed is "raw", but the result itself is already HTML-formatted. Let's say you're requesting for a course entity. Instead of getting the entity itself, you'll get something like
```html
<div>
    <p>This is the course name</p>
    <p>This is the course ID</p>
    <p>This is the course ...</p>
</div>
```
i.e. the data has been pre-processed in the back-end. So saying the result is "raw" is not entirely accurate. Yet saying it "not raw" will also draw debates 😛